### DISCUSSION TITLE
Angular bindings

### DISCUSSION THREAD

**Discussion by nhducit** [TanStack/query#3694] [Ideas]
hi everyone our company has several Angular projects and looking for a better way to handle server cache. React-query is always the library I wished it could exist for angular. Now, v4 is already providing bindings for React, Vue, and Solid. I really wish that we can also make Angular bindings.

I'm exploring query core to make the Angular bindings, I have some questions:
- Do we have any documents to create a binding?
- Are there any recommended approaches I should follow?

As far as I know, query core already provides the caching/invalidate cache mechanism, What I need to do are
- set up angular listen to query data
- unsubscribe query if the angular component is unmounted

Someone implement a library with similar ideas but it's deprecated now
https://github.com/timdeschryver/rx-query

Update 1:
- due to the Angular rendering nature, the Angular Class component is only executed for the first time. So we need to find a way to notify the created query-dependent array is updated and the query also needs to notify the updated query result to the component. rxjs Subject (maybe BehaviorSubject? I'm not an rxjs expert) seems to be a good candidate here.

> A BehaviorSubject holds one value. When it is subscribed it emits the value immediately. A Subject doesn't hold a value.

We should use BehaviorSubject instead of Subject

```
class AngularComponent() {

users$ = useQuery(queryDependentArray$, queryFn)
}
```

the dollar sign $ is a naming convention that marks the variable as an rxjs Observable

**Comment by andrey99z** [comment]
Same here, need replacement for rx-query and react-query seems perfect with it's core not being react dependent.

I've got some working code here:
https://codesandbox.io/s/sad-jones-km6tq9

Not sure if this is the right approach, happy to discuss.

**Reply by nhducit** [reply]
hi @andrey99z great. I will take a look at your example

here is my naive approach, I'm exploring making the useQuery's params an observable

[stackblitz.com/edit/angular-ivy-j8pqww?file=src%2Fapp%2Fapp.component.ts](https://stackblitz.com/edit/angular-ivy-j8pqww?file=src%2Fapp%2Fapp.component.ts)

**Reply by nhducit** [reply]
hi @andrey99z could you create a github repo to discuss easier?

Why do we need to wrap the result in `defer`? I found that rx-query is also using the same technique
```
  const deferred = defer(() => {
    let observer: any;
    let unsubscribe: any;
```

**Reply by nhducit** [reply]
you might already knew this 🙏 we can create new Github repo directly from codesandbox

<img width="1140" alt="image" src="https://user-images.githubusercontent.com/4246176/173283739-c47ad996-d692-4e86-a521-496b31dff211.png">

**Reply by nhducit** [reply]
my 2 cents: it would be great if you can add a comment to explain why do we need to call some react-query core APIs. Because not everyone is familiar with the react query core API 🙏

**Reply by andrey99z** [reply]
https://github.com/andrey99z/ng-query
I added comments and sent you the invite

**Reply by andrey99z** [reply]
I had a look at your sandbox, looks like you doing it in a very similar way!

**Reply by nhducit** [reply]
nice, thanks @andrey99z I'm following React bindings implementation, read rx-query source code

**Reply by nhducit** [reply]
I created a new issue FYI
https://github.com/andrey99z/ng-query/issues/1

**Reply by andrey99z** [reply]
Great, suggest moving further discussions there.

**Reply by ApplY3D** [reply]
Hi guys, Im interested too about `react-query` for Angular 🔥

**Reply by andrey99z** [reply]
@ApplY3D your contributions are welcome :)

**Comment by trungvose** [comment]
The ngneat team from @NetanelBasal has just released a beta version for `ngneat/query`. Check it out guys 👋

https://github.com/ngneat/query

**Reply by ApplY3D** [reply]
react deps instead of direct core [@tanstack/query-core](https://www.npmjs.com/package/@tanstack/query-core) dep only while beta?

**Comment by rbalet** [comment]
TanStack Query now has its own Angular library! 🎉
Feel free to try it out: [TanStack Query for Angular](https://tanstack.com/query/latest/docs/framework/angular/overview).
